### PR TITLE
Check whether fixed point decimal places fit into total bits

### LINKED ISSUE #4061: Fixed point type fractional digit limit
Currently the limit is set at maximum 80 digits, that means the largest number representable would be
`0.999999999999999999999999999999999999999999999999999999999999999999999999999999999` in the `ufixed256x80` type, while the largest number fitting 256 bits is `115792089237316195423570985008687907853269984665640564039457584007913129639936`.

`999999999999999999999999999999999999999999999999999999999999999999999999999999999`
is larger than
`115792089237316195423570985008687907853269984665640564039457584007913129639936`

It actually requires 266 bits (`log2(pow(10, 80) - 1)`). I think the proper limit is 77 digits.

Not to mention signed fixed point, where an extra bit must be reserved for the sign.

### PR DESCRIPTION
**Pull request by axic**
Fixes #4061.

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -604,6 +601,12 @@ FixedPointType::FixedPointType(int _totalBits, int _fractionalDigits, FixedPoint
 		"Invalid bit number(s) for fixed type: " + 
 		dev::toString(_totalBits) + "x" + dev::toString(_fractionalDigits)
 	);
+
+	unsigned const bitsAvailable = m_totalBits - (isSigned() ? 1 : 0);
+	solAssert(
+		fitsPrecisionExp(10, m_fractionalDigits, bitsAvailable),
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: This should actually become a static helper and be used in the typechecker too, to restrict invalid combinations:
```
static FixedPointType::isValid(unsigned _totalBits, unsigned _fractionalDigits, bool _isSigned)
{
    unsigned const bitsAvailable = _totalBits - (_isSigned ? 1 : 0);
    return fitsPrecisionExp(10, _fractionalDigits, bitsAvailable);
}
```

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -665,15 +668,15 @@ string FixedPointType::toString(bool) const
 bigint FixedPointType::maxIntegerValue() const
 {
 	bigint maxValue = (bigint(1) << (m_totalBits - (isSigned() ? 1 : 0))) - 1;
-	return maxValue / pow(bigint(10), m_fractionalDigits);
+	return maxValue / boost::multiprecision::pow(bigint(10), m_fractionalDigits);
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: This case didn't had a `using` statement, so not sure what `pow` did it pick up.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -665,15 +668,15 @@ string FixedPointType::toString(bool) const
...
 }
 
 bigint FixedPointType::minIntegerValue() const
 {
 	if (isSigned())
 	{
 		bigint minValue = -(bigint(1) << (m_totalBits - (isSigned() ? 1 : 0)));
-		return minValue / pow(bigint(10), m_fractionalDigits);
+		return minValue / boost::multiprecision::pow(bigint(10), m_fractionalDigits);
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: Same here.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -1138,7 +1138,7 @@ u256 RationalNumberType::literalValue(Literal const*) const
 		auto fixed = fixedPointType();
 		solAssert(fixed, "");
 		int fractionalDigits = fixed->fractionalDigits();
-		shiftedValue = m_value.numerator() * pow(bigint(10), fractionalDigits) / m_value.denominator();
+		shiftedValue = m_value.numerator() * boost::multiprecision::pow(bigint(10), fractionalDigits) / m_value.denominator();
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: And here.

**Comment by axic** [comment]
@chriseth can you have a high level look at this and the referenced issue?

**Comment by axic** [comment]
@chriseth we should discuss the general motivation of this latest on Wednesday so it can be finished.

File: libsolidity/ast/Types.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: This was argument-dependent lookup (https://en.cppreference.com/w/cpp/language/adl) - since ``bigint`` is in ``boost::multiprecision``, the ``pow`` of ``boost::multiprecision`` will be considered during lookup (and that's probably the only ``pow`` that can deal with ``bigint``s). But it's probably better to add the namespace explicitly - argument-dependent lookup is a somewhat weird concept...

File: libsolidity/ast/Types.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Actually I think it may be better to check, whether 10^m_fractionalDigits <= 2^bitsAvailable directly here. The motivation behind ``fitsPrecisionExp`` was to restrict exponentiation of literals to **roughly** 4096 bits - and we didn't want to do the exponentiation before the check, since calculating it could already have eaten up a lot of memory. However, the calculation is approximative (maybe that should be documented...). That's not a problem for literals (the idea was to cut off at some random insanely high value that's *around* 2^4096), but here (1) precision does matter and (2) ``m_fractionalDigits`` and ``m_totalBIts`` (resp. ``bitsAvailable``) is limited to reasonably low values anyways, so doing the precise calculation doesn't hurt.

Comment (axic) [reply libsolidity/ast/Types.cpp]: I would even say that we could just pre-calculate the max digits supported for every bit width between 8 and 256 in 8 bit increments.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Well, we could do that... but it may be more difficult to see what's going on and whether the values are correct that way. I would say something like ``boost::multiprecision::exp(bigint(10), m_fractionalDigits) <= boost::multiprecision::exp(bigint(2), bitsAvailable)`` doesn't hurt too much.

Actually, in general I think it's rather weird that fixed byte types use a fixed number of decimal digits for the fractional part instead of a fixed fractional bit resolution, but judging from the related issues it seems you had discussed and decided that before my time :).

Comment (axic) [reply libsolidity/ast/Types.cpp]: It was fractional bits before, but it seems to be more natural to know number of decimal digits in calculations as a user.

It is more painful in some other parts from a tooling perspective.

Comment (axic) [reply libsolidity/ast/Types.cpp]: Ah I remember the limitation was that `exp` was not available in the boost we're using. Any suggestion? Well actually there's `pow2`, but no `pow10`.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Ah, ok. Actually, for that I would have used a table in the docs that tells you the number of decimal digits a given amount of fractional bits gets you :). But if users want it this way, then we should of course do it this way (even though I imagine that not only tooling will be more painful, but conversions and arithmetic using fixed point numbers might also cost more gas this way in the end).

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Ehm, sorry, I think it should be ``pow``, not ``exp``, but that should be in boost.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: So either ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= boost::multiprecision::pow(bigint(2), bitsAvailable)``
or ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= boost::multiprecision::pow2(bitsAvailable)``
or even ``boost::multiprecision::pow(bigint(10), m_fractionalDigits) <= (bigint(1) << bitsAvailable)`` should all be fine.

**Comment by axic** [comment]
@ekpyron reworked, please review. Probably also should squash the last two commits into one.

File: libsolidity/ast/Types.cpp [hunk 5]
```diff
@@ -617,6 +617,17 @@ FixedPointType::FixedPointType(unsigned _totalBits, unsigned _fractionalDigits,
...
 		dev::toString(_totalBits) + "x" + dev::toString(_fractionalDigits)
 	);
+
+	solAssert(
+		isValid(_totalBits, _fractionalDigits, isSigned()),
+		"Number of fractional digits do not fit the specified bit width"
+	);
+}
+
+bool FixedPointType::isValid(unsigned _totalBits, unsigned _fractionalDigits, bool _isSigned)
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Maybe this should also involve the lower and upper bounds for total bits, etc., used in the constructor of ``FixedPointType``? Having a function called ``isValid`` defined this way somewhat suggests that ``fixed512x1`` or ``fixed7x1`` would be valid... if this should still only check whether the fractional part fits into the bit width, maybe the function should be renamed accordingly?

File: test/libsolidity/SolidityTypes.cpp [hunk 6]
```diff
@@ -61,6 +61,10 @@ BOOST_AUTO_TEST_CASE(fixed_types)
 {
 	BOOST_CHECK(*Type::fromElementaryTypeName(ElementaryTypeNameToken(Token::Fixed, 0, 0)) == *make_shared<FixedPointType>(128, 18, FixedPointType::Modifier::Signed));
 	BOOST_CHECK(*Type::fromElementaryTypeName(ElementaryTypeNameToken(Token::UFixed, 0, 0)) == *make_shared<FixedPointType>(128, 18, FixedPointType::Modifier::Unsigned));
+
+	/// @todo add more comprehensive tests
+	BOOST_CHECK_EQUAL(FixedPointType::isValid(256, 77, 0), true);
```
Comment (ekpyron) [inline test/libsolidity/SolidityTypes.cpp]: As the comment above says, there should indeed be some more tests about more corner cases at some point. However, for now: why use ``0`` and not ``false`` as third argument? And maybe it makes sense to already add tests for the signed 256-bit case as a reminder to also test the signed case more extensively later on.

**Review by ekpyron — COMMENTED** [review]
Looks fine to me now, apart from the minor comments.

File: test/libsolidity/SolidityTypes.cpp (hunk 6 above)
Comment (axic) [reply test/libsolidity/SolidityTypes.cpp]: Good catch about `0`.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/ast/Types.cpp]: Good idea also.

File: test/libsolidity/SolidityTypes.cpp (hunk 6 above)
Comment (axic) [reply test/libsolidity/SolidityTypes.cpp]: Changed and added more tests. The majority of this should be handled in #3389 where `isValid` should also be used in the `TypeChecker` to show proper errors.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/ast/Types.cpp]: Added, though perhaps in #3389 the `TypeChecker` may have two helpers ("out of range" or "invalid configuration").

File: libsolidity/ast/Types.cpp [hunk 7]
```diff
@@ -617,6 +617,20 @@ FixedPointType::FixedPointType(unsigned _totalBits, unsigned _fractionalDigits,
 		"Invalid bit number(s) for fixed type: " +
 		dev::toString(_totalBits) + "x" + dev::toString(_fractionalDigits)
```
Comment (axic) [inline libsolidity/ast/Types.cpp:618]: I left this here for better "error reporting" for the moment.

**Comment by ekpyron** [comment]
The test failures are interesting - I think ``RationalNumberType::fixedPointType()`` may have a problem...

**Comment by axic** [comment]
I think the `type_identifiers` test fails on the invalid `ufixed8x64`. Need to investigate `RationalNumberType`.

**Comment by ekpyron** [comment]
``312_leading_zero_rationals_convert`` fails, because ``RationalNumberType::fixedPointType()`` tries to create an invalid ``fixed128x52``, so something needs to be done there.
However, I'm also noticing, that ``soltest`` for some reason doesn't seem to properly report the exceptions for the failing ``solAssert``s - I'll open an issue for that and will look into it.

**Comment by axic** [comment]
The bug is in `RationalNumberType::fixedPointType()`. Just trying a fix now.

**Comment by axic** [comment]
Some test expectations need to updated, but in some cases the calculation seems to be off.

For example in `syntaxTests/nameAndTypeResolution/312_leading_zero_rationals_convert.sol` it deduces:

`fixed128x52` for the literal `0.0000000000000006661338147750939242541790008544921875` (and tries to assign it to `fixed256x52`).

The `totalBits` part of the deduction is wrong, it won't fit 52 decimal digits.

**Comment by axic** [comment]
The problem with the calculation is that the leading zeroes are ignores and `6661338147750939242541790008544921875` does fit into 128 bits.

File: libsolidity/ast/Types.cpp [hunk 8]
```diff
@@ -617,6 +617,21 @@ FixedPointType::FixedPointType(unsigned _totalBits, unsigned _fractionalDigits,
...
+	);
+}
+
+bool FixedPointType::isValid(unsigned _totalBits, unsigned _fractionalDigits, bool _isSigned)
+{
+	if (_totalBits < 8 || _totalBits > 256 || _totalBits % 8 != 0 || _fractionalDigits > 80)
+		return false;
+
+	unsigned const bitsAvailable = _totalBits - (_isSigned ? 1 : 0);
+	return boost::multiprecision::pow(bigint(10), _fractionalDigits) <= (bigint(1) << bitsAvailable);
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp:634]: What exactly does this check? That `1.0` is representable? Why is it `<=` and not `<`?

Comment (chriseth) [reply libsolidity/ast/Types.cpp:634]: @axic @meowingtwurtle

Comment (ekpyron) [reply libsolidity/ast/Types.cpp:634]: @chriseth I think it checks whether the number of fractional decimal digits is small enough to actually fit into the total number of bits.
The idea is that a type like ``fixed1x2`` doesn't make much sense or is at least confusing, since two decimal digits can't be represented in one bit. If ``10^_fractionalDigits <= 2^bitsAvailable``, then at least the number of decimal digits supposed to be representable in the type is actually representable in the type.

If, on the other hand, you think of ``ufixedXxY`` as a type with ``X`` bits that is implicitly multiplied by ``10^(-Y)``, then this is the same as checking whether ``1.0`` is representable.
Arguably one could do the opposite and allow a ``fixed1x2`` type, if one thinks about them this way. To be honest I find the current naming of fixed point values as ``X`` being total bits and ``Y`` being decimal places pretty strange in general, regardless of how one deals with cases like this.

In any case: if one wants this check, I think ``<=`` makes more sense than ``<``.

Comment (randomnetcat) [reply libsolidity/ast/Types.cpp:634]: Well it is mathematically impossible for them to be equal, so it doesn't really matter at all.

Proof: 10 is divisible by 5, so (10^n1) is also divisible by five. 2 is not divisible by 5, so (2^n2) is not divisible by 5. It is impossible for a number divisible by 5 to be equal to one that is not divisible by 5. QED.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp:634]: Actually, ``<=`` and ``<`` will even be the same in this case, since ``10^x`` will never actually be equal to ``2^y``, unless ``x`` and ``y`` are zero and ``_totalBits`` is already required to be at least ``8``.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp:634]: @meowingtwurtle Yep, that's what I just noticed as well :).

Comment (chriseth) [reply libsolidity/ast/Types.cpp:634]: Could you please add a comment? I think it is better to explain it using `1.0 is representable`, since the sentence `whether the number of fractional decimal digits is small enough to actually fit into the total number of bits.` still lacks a definition of what it means for fractional digits to fit into the total number of bits.

**Comment by randomnetcat** [comment]
Proposal to calculate `totalBits`

To determine bits needed:
```cpp
rational const absValue = abs(m_value);
// If there is no integer part, add 1 to create an integer part, so we have all of the digits used for totalBits calculation.
rational const scaledValue = (absValue + (absValue < 1 ? 1 : 0)) * pow(bigint(10), fractionalDigits);
bigint const truncatedValue = (negative ? -1 : 1) * (scaledValue.numerator() / scaledValue.denominator());
unsigned totalBits const = max(bytesRequired(truncatedValue), 1u) * 8;
```

@axic

**Comment by chriseth** [comment]
Closing just as https://github.com/ethereum/solidity/pull/4980
